src/dct/assets/AssetManager.lua

```diff
@@ -149,21 +149,21 @@ function AssetManager:getAssetByDCSObject(dcsObjName)
 end
 
 --[[
--- filterAssets - return all asset names matching `filter`
--- filter(asset)
---   returns true if the filter matches and the asset name should be kept
--- Return: a table with asset names as keys. Will always returns a table,
---   even if it is empty
+-- filterAssets - return all assets matching filter function
+-- arg: filter
+--   Return a truthy value for assets that should be returned by this function
+-- Returns a list of matched assets, even if empty
 --]]
 function AssetManager:filterAssets(filter)
 	checklib.func(filter)
 
 	local list = {}
 	for name, asset in pairs(self._assetset) do
 		if filter(asset) then
-			list[name] = true
+			list[name] = asset
```

Comment:
why? Oh I see instead of having the caller now loop back over that list to gain access to the asset by name you want to provide the ability to just use the list returned. This is fine as long as the caller doesn't hold on to this list. What we could do is make `list` have values that are weak references[1]. That way the asset manager could still be considered to have the single list of asset references. Then the caller could just do want they need with the list and even if they store it off somewhere the references to the assets in the tables will not count against garbage collection and the Asset object will be collected when AssetManager deletes the Asset from its asset list.

[1] https://www.lua.org/pil/17.html

Reply:
That's an interesting idea, but the it relies on the non-deterministic GC, and the caller can still trivially create a strong reference by copying it to a new, non-weak table. i.e.

```lua
local assets = assetmgr:filterAssets(filter)
local myList = { assets[1] }
-- myList can now hold assets[1] in memory indefinitely
```

And even if `filterAssets` didn't yield asset references, `getAsset` would still hand them over on request. The best we can do is to manually avoid breaking invariants (like trying to spawn a dead asset), as we can't trust the GC to nil out a value in time.